test/rules/unrelated_type_equality_checks_test.dart

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2023, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:test_reflective_loader/test_reflective_loader.dart';
+
+import '../rule_test_support.dart';
+
+main() {
+  defineReflectiveSuite(() {
+    defineReflectiveTests(UnrelatedTypeEqualityChecksTestLanguage300);
+  });
+}
+
+@reflectiveTest
+class UnrelatedTypeEqualityChecksTestLanguage300 extends LintRuleTest
+    with LanguageVersion300Mixin {
+  @override
+  bool get dumpAstOnFailures => true;
+
+  @override
+  String get lintRule => 'unrelated_type_equality_checks';
+
+  test_switchExpression() async {
+    await assertDiagnostics(r'''    
+const space = 32;
+
+String f(int char) {
+  return switch (char) {
+    == 'space' => 'space',
+  };
+}
+''', [
+      error(CompileTimeErrorCode.NON_EXHAUSTIVE_SWITCH, 49, 6),
```

Comment:
You can, and really should, get rid of extraneous errors. While we make an effort for the lints to work well in the face of errors, that isn't really what we're trying to test here. In most cases it's as simple as converting to use a switch statement rather than an expression. For `test_switchExpression_lessEq_ok` you can probably just reverse the types of the param and value.

Reply:
Thanks!  I'll do a round of cleanup.